ubuntu_image/common_builder.py

```diff
@@ -349,6 +349,23 @@ def prepare_filesystems(self):
         self._next.append(self.populate_filesystems)
 
     def _populate_one_volume(self, name, volume):
+        # For the LK bootloader we need to copy boot.img and snapbootsel.bin to
+        # the gadget folder so they can be used as partition content. The first
+        # one comes from the kernel snap, while the second one is modified by
+        # 'snap prepare-image' to set the right core and kernel for the kernel
+        # command line.
+        if volume.bootloader is BootLoader.lk:
```

Comment:
A few things here. I'm wondering if this is the best fit for the boot copies to gadget actually. We do something similar-ish _populate_one_bootfs() - since basically this is part of preparation of the bootfs, right? Is there any reason why we do it in _populate_one_volume instead (which is more for population of non-boot partitions)?

The other thing is that I guess at first I was a bit concerned about this being run per-volume. Since we run this for every volume defined in the gadget, so in theory, conceptually, all the copies should be done per-volume. But on the other hand, I doubt we'll actually have any devices that would have 2+ volumes from which 2 have BootLoader.lk specified - there doesn't seem to be any practical reasons for that. So I guess it's all good from this POV.

Reply:
I do not think we can do this in `_populate_one_bootfs()`, as what is done there is populating a partition with files. In this case we need image files that are directly the partition content. And those files do not come from the gadget:
* boot.img comes from the kernel snap - we need it there so kernel updates can be performed
* snapbootsel.bin is modified by 'snap prepare', similarly to what is done to uboot/grub environment files - we need to take it after that has happened

Re: per-volume, I agree, I do not think we would have a case with more than one volume with lk.

Reply:
Ah! Ok, that makes much more sense, this wasn't entirely clear to me without seeing a gadget.yaml - now looking at the one in the test-case below, I finally understand how this works. My original understanding was that it was files that had to be present on the filesystem for those partitions.

So in this case I think that conceptually this change basically doesn't really fit in either - so let's leave it as is. It could be a potential good fit for _populate_one_bootfs() because it's basically 'preparing' boot-related files from the gadget for future copying over into the image, but it also fits _populate_one_volume() because that's where the binary copying is happening. Anyway, it's good as is, no need to bikeshed!